### ISSUE TITLE
Improve travis speed again

### ISSUE THREAD

**Issue by diox** [mozilla/addons-server#2228]
We splitted travis jobs in #2112 but travis seems to limit builds to 5 concurrent jobs. Thankfully, 3 of our jobs (flake8, assets and docs) are fairly simple and are over quickly, but maybe we can speed things up more.

One idea: I noticed we install ES in `.travis.yml` in `before_install`, but if we only use it in a subset of the tests (in fact, we should only need it for the `es` job), we should be able to move the install to the job dealing with that subset, instead of slowing down the install part of each job ?

**Comment by diox** [comment]
Regarding elasticsearch: it takes on average one second, so not really worth pursuing.

Python package install takes more time. We could try caching things:
```
cache:
  directories:
    - node_modules
    - .tox
    - $HOME/.cache/pip/
```

And maybe not re-create tox everytime ? Not sure.

**Comment by diox** [comment]
Random counter-intuitive idea: if speeding the assets job up is not enough, bundling docs and flake8 together in a single build could improve things, because the last job would then be able to start sooner.

**Comment by EnTeQuAk** [comment]
I just noticed that we're not downloading product-details on travis, see https://github.com/mozilla/addons-server/blob/master/settings_test.py#L148 and https://github.com/mozilla/addons-server/blob/master/src/olympia/core/apps.py#L53

**Comment by diox** [comment]
assets build still downloads it: https://travis-ci.org/mozilla/addons-server/jobs/129709236

**Comment by diox** [comment]
#2652 fixed the `assets` job so that it no longer downloads product details. We seemed to have gained a couple minutes there. I feel improving that job is still the way to go to make the overall build faster. Obviously reworking our entire assets pipeline is out of the question :) A couple ideas though:
- Reduce number of assets. We probably still reference files that no longer exist, things like that
- Make the job happen in a in-memory filesystem or something like that ? I don't know if that's already the case or if it's possible.

**Comment by diox** [comment]
Other idea: `collectstatic` is copying a bunch of files we don't need to be copied in the final static dir ; we should only need the final bundles and the images.

**Comment by andymckay** [comment]
Is improving this still an issue? Seems pretty fast to me these days. Perhaps I'm used to tryserver.

**Comment by diox** [comment]
It's still not amazing, but we've exhausted the easy quick wins I think. Let's close for now.

**Comment by EnTeQuAk** [comment]
Alrighty, I just spend a hour splitting up tasks and moving dependency installation to travis to get proper timings and maybe see more bottlenecks. See https://travis-ci.org/mozilla/addons-server/jobs/140719163 as a very very good example. This runs most of our test-suite, the suite itself runs in under a minute (https://travis-ci.org/mozilla/addons-server/jobs/140719163#L709) but dependency installation takes 163 seconds (https://travis-ci.org/mozilla/addons-server/jobs/140719163#L302)

Now, the problem is that we can't improve much of it. It's mostly slow because we have to build wheel packages for quite a lot of packages on every build (e.g pillow, lxml but even smaller packages like django-tables2 and others) since they either don't provide proper linux wheels but only mac and windows or none at all.

Now one might think, let's build them with `pip wheel`, cache them and be done with it. That would probably speed up tests by minutes but we cannot do that because of our usage of hashes :cry: which will fail for our built packages. This is probably also the reason why the caching of `$HOME/.cache/pip` doesn't help too much.

tldr; it sucks and is hard to impossible to improve currently imho

**Comment by EnTeQuAk** [comment]
In case some one is interested, statistics of what module has how many tests:

* abuse:  5
* access:  24
* accounts:  111
* addons:  818
* amo:  491
* api:  127
* applications:  10
* bandwagon:  170
* blocklist:  75
* browse:  121
* compat:  30
* devhub:  868
* discovery:  5
* editors:  635
* files:  348
* internal_tools:  25
* landfill:  50
* legacy_api:  108
* legacy_discovery:  51
* lib:  71
* pages:  6
* paypal:  31
* reviews:  94
* search:  182
* signing:  39
* stats:  129
* tags:  38
* translations:  87
* users:  331
* versions:  208
* zadmin:  17

**Comment by EnTeQuAk** [comment]
More interesting facts: The same build amount of tests with different cores:

 * 2 cores: 359.36s (https://travis-ci.org/mozilla/addons-server/jobs/140746439)
 * 3 cores: 539.11s (https://travis-ci.org/mozilla/addons-server/jobs/140746499)

**Comment by EnTeQuAk** [comment]
Shaved off another ~3 minutes to master: https://travis-ci.org/mozilla/addons-server/builds/140748773 and if we make elasticsearch tests multicore-compatible we might be able to get <10 min: https://travis-ci.org/mozilla/addons-server/builds/140752083 (and maybe split off `zadmin` but that needs some benchmarks)
